### PR TITLE
Add Persistent storage integration for scoreboard with Firebase

### PR DESCRIPTION
**Pull request by mihai-iacov**
This adds support for Firebase realtime DB - documents are saved in JSON format in a Firebase DB cluster (west-europe).

CC: @raduciobanu, @radumarin88, for tracking the project progress

Sample app view:
Screenshot 2022-01-14 at 13 39 31

Sample FirebaseDB view:
Screenshot 2022-01-14 at 13 40 40

### CODE REVIEW

File: app/src/main/java/com/example/sudoku/ScoreboardEntryModel.kt [hunk 1]
```diff
@@ -1,4 +1,15 @@
 package com.example.sudoku
 
-data class ScoreboardEntryModel(val rank: String, val playerName: String, val score: String) {
+import com.google.firebase.database.IgnoreExtraProperties
+
+@IgnoreExtraProperties
+data class ScoreboardEntryModel(val rank: String = "", val playerName: String = "", val score: String = "") {
```
Comment (vmdamian) [inline app/src/main/java/com/example/sudoku/ScoreboardEntryModel.kt:6]: I would add an ID here for the device, so in the way we configured it, the user can register himself in the scoreboard with whatever nickname he wants. If we add the ID on top of that, we can highlight hI would add an ID here for the device, so in the way we configured it, the user can register himself in the scoreboard with whatever nickname he wants. If we add the ID on top of that, we can highlight his score entries on the scoreboard, no mattering what nickname he used. A somewhat good ID I think could be a hash on the MAC address of the device. It's not perfect, but good enough I think.is score entries in the scoreboard, not mattering what the nickname he used. A somewhat good ID I think could be a hash on the MAC address of the device. It's not perfect, but good enough I think.

**Review by vmdamian — COMMENTED** [review]
The "Submit highscore" button is visible even if I access the scoreboard from main menu. That doesn't really make sense. I think there should be 2 ways of accessing the scoreboard:
- from the win screen, where we can put the nickname input and submit highscore button, which then transitiones to the scoreboard fragment
- from the main menu

**Comment by vmdamian** [comment]
We should make a decision around implementing/deleting the best times tab in the scoreboard.

**Review by vmdamian — APPROVED** [review]

**Comment by mihai-iacov** [comment]
> We should make a decision around implementing/deleting the best times tab in the scoreboard.

I would say - let's drop the "best times" tab for now, to reduce the current scope. The high priority feature is presented regardless of it (download to and upload from the remote storage DB).

File: app/src/main/java/com/example/sudoku/ScoreboardEntryModel.kt (hunk 1 above)
Comment (mihai-iacov) [reply app/src/main/java/com/example/sudoku/ScoreboardEntryModel.kt:6]: The main idea sounds nice and solid, but it makes the implementation more complex. I would delay it for another PR and see if we have enough time to add it to the project.
Until then, I think we can ask the user for a nickname every time when they submit new high scores (like in old arcade games).

**Comment by mihai-iacov** [comment]
> The "Submit highscore" button is visible even if I access the scoreboard from main menu. That doesn't really make sense. I think there should be 2 ways of accessing the scoreboard:
> 
> * from the win screen, where we can put the nickname input and submit highscore button, which then transitiones to the scoreboard fragment
> * from the main menu

I updated the functionality:
- the highscores screen only displays the current leaderboard (read-only)
- the win screen allows you to add your nickname and submit the highscore; if you submit the score, then the app navigates to the scoreboard

**Pull request merged**